src/commands/Changelog.ts

```diff
@@ -0,0 +1,318 @@
+ */
+export class Changelog {
+    private changelog = new Map<string, ChangelogEntry>();
+
+    /**
+     * Gets the changes for one node type.
+     * @param nodeType
+     */
+    get(nodeType: string) {
+        return this.changelog.get(nodeType);
+    }
+
+    /**
+     * Gets access to the map of all changelog entries.
+     */
+    getAll() {
+        return this.changelog;
+    }
+
+    /**
+     * Clear the changelog.
+     */
+    clear() {
+        this.changelog.clear();
+    }
+
+    /**
+     * Logs that an expression was added to a language.
+     * @param expression The expression.
+     * @param languageAbbr Abbreviation of the language.
+     */
+    addedNodeToLanguage(expression: ExpressionMetricMapping, languageAbbr: string) {
+        // There is no changelog entry for this expression yet, so create it.
+        if (!this.changelog.has(expression.expression)) {
+            this.changelog.set(
+                expression.expression,
+                new ChangelogEntry(
+                    expression.expression,
+                    false,
+                    expression.languages,
+                    [],
+                    [],
+                    expression.languages
+                )
+            );
+        }
+        this.changelog.get(expression.expression)?.addedToLanguage(languageAbbr);
+    }
+
+    /**
+     * Logs that a totally new node type/expression was added to a language.
+     * @param nodeTypeName Name of the node type.
+     * @param languageAbbr Abbreviation of the language to which this node was added.
+     */
+    addedNewNode(nodeTypeName: string, languageAbbr: string) {
+        const entry = new ChangelogEntry(nodeTypeName, true);
+        entry.addedToLanguage(languageAbbr);
+        this.changelog.set(nodeTypeName, entry);
+    }
+
+    /**
+     * Logs that an expression was removed from a language.
+     * @param expression The expression.
+     * @param languageAbbr Abbreviation of the language.
+     */
+    removedNodeFromLanguage(expression: ExpressionMetricMapping, languageAbbr: string) {
+        if (!this.changelog.has(expression.expression)) {
+            this.changelog.set(
+                expression.expression,
+                new ChangelogEntry(
+                    expression.expression,
+                    false,
+                    expression.languages,
+                    [],
+                    [],
+                    expression.languages
+                )
+            );
+        }
+        this.changelog.get(expression.expression)?.removedFromLanguage(languageAbbr);
+    }
+
+    /**
+     * Writes the changelog.
+     * @param metricMappings The current mappings to inform the user about which metric(s)
+     * are currently calculated with the changed syntax node(s).
+     * @return Promise that fulfills once the changelog has been written successfully.
+     */
+    writeChangelog(metricMappings: Map<string, ExpressionMetricMapping>) {
```

Comment:
Let's split this function a little bit to be more readable.

Reply:
Splitted into three new functions